hugr-model/src/v0/text/parse.rs

```diff
@@ -0,0 +1,716 @@
+            targets: &[],
+            children,
+            meta,
+            r#type,
+        });
+
+        self.module.root = root_region;
+
+        Ok(())
+    }
+
+    fn parse_term(&mut self, pair: Pair<'a, Rule>) -> ParseResult<TermId> {
+        debug_assert!(matches!(pair.as_rule(), Rule::term));
+        let pair = pair.into_inner().next().unwrap();
+        let rule = pair.as_rule();
+        let mut inner = pair.into_inner();
+
+        let term = match rule {
+            Rule::term_wildcard => Term::Wildcard,
+            Rule::term_type => Term::Type,
+            Rule::term_static => Term::StaticType,
+            Rule::term_constraint => Term::Constraint,
+            Rule::term_str_type => Term::StrType,
+            Rule::term_nat_type => Term::NatType,
+            Rule::term_ctrl_type => Term::ControlType,
+            Rule::term_ext_set_type => Term::ExtSetType,
+
+            Rule::term_var => {
+                let name_token = inner.next().unwrap();
+                let name = name_token.as_str();
+                Term::Var(LocalRef::Named(name))
+            }
+
+            Rule::term_apply => {
+                let name = GlobalRef::Named(self.parse_symbol(&mut inner)?);
+                let mut args = Vec::new();
+
+                for token in inner {
+                    args.push(self.parse_term(token)?);
+                }
+
+                Term::Apply {
+                    name,
+                    args: self.bump.alloc_slice_copy(&args),
+                }
+            }
+
+            Rule::term_apply_full => {
+                let name = GlobalRef::Named(self.parse_symbol(&mut inner)?);
+                let mut args = Vec::new();
+
+                for token in inner {
+                    args.push(self.parse_term(token)?);
+                }
+
+                Term::ApplyFull {
+                    name,
+                    args: self.bump.alloc_slice_copy(&args),
+                }
+            }
+
+            Rule::term_quote => {
+                let r#type = self.parse_term(inner.next().unwrap())?;
+                Term::Quote { r#type }
+            }
+
+            Rule::term_list => {
+                let mut items = Vec::new();
+                let mut tail = None;
+
+                for token in filter_rule(&mut inner, Rule::term) {
+                    items.push(self.parse_term(token)?);
+                }
+
+                if inner.next().is_some() {
+                    let token = inner.next().unwrap();
+                    tail = Some(self.parse_term(token)?);
+                }
+
+                Term::List {
+                    items: self.bump.alloc_slice_copy(&items),
+                    tail,
+                }
+            }
+
+            Rule::term_list_type => {
+                let item_type = self.parse_term(inner.next().unwrap())?;
+                Term::ListType { item_type }
+            }
+
+            Rule::term_str => {
+                // TODO: Escaping?
```

Comment:
Also here.

Reply:
Also postponed to https://github.com/CQCL/hugr/issues/1549.